Release process proposal (#91)
kyma-project/test-infra: docs/kyma-release-process.md

For every component, there is a presubmit job for release branches, that is responsible for publishing docker images with release tag (eg. 0.4.3). 
```
    name: prow/kyma/components/ui-api-layer/release
    branches:
      - release-X.Y
    run_if_changed: ...values.yaml,components-ui-api-layer
```
Calculation of proper image tag is described in section "Calculation release image tag" [here](#calculation-release-image-tag)

Such configuration implies that after successful modification of all values.yaml files, all images are published.
There can be an additional job, that checks if in all `values.yaml` all components use exactly the same version (eq. 0.4.3).
For release branches and master branch, branch protection rules needs to be defined, that mark all checks as a required. Without that, merge button is enabled
even if some checks failed or are in progress. 

3. Integration jobs needs to be executed. There are following options how to configure them:
- run it on every change on PR. This is very simple solution from Prow perspective. Unfortunately, there is a high possibility that this job will fail at the beginning. 

>  This job should be executed only after all all 

Author: to discuss, IMO meaning of sentence has changed.

that do not pass integration tests.

4. Publishing artifacts and release creation. There are following options how to configure them:
- run jobs after integrations jobs:
```
    name: prow/kyma/integration
    branches:
      - release-X.Y
    // run integration tests
    run_after_success:
      - build kyma installer
      - publish artifacts
      - create release
```
 This can be configured as a Presubmit (on PR) or Postsubmit job. In case of Presubmit job, it is easiliy to retrigger such action
 by adding proper command in PR. In case of Postsubmit job, such option does not exist. 
 - trigger jobs by specifying comments on PR. This can be an option, if there are some manual steps required before publishing artifacts. 

### Consequences
In proposed approach, we have to **define additional Presubmit job for every component responsible for releasing component.** 
Such job is almost the same as Presubmit job for master branch, except:
- branch name
- run_if_changed parameters
- different labels: preset-build-release instead of preset-build-pr

> To reduce boilerplate code and code repetition, we can use yaml features, such as extending objects.

Reviewer: ```suggestion
To reduce boilerplate code and code repetition, use `yaml` features, such as extending objects.
```

    repo: test-infra
    base_ref: master
    path_alias: github.com/kyma-project/test-infra
  spec:
    containers:
    - image: eu.gcr.io/kyma-project/prow/buildpack-golang:0.0.1
      securityContext:
        privileged: true
      command:
      - "/home/prow/go/src/github.com/kyma-project/test-infra/prow/scripts/pipeline.sh"
      env:
      - name: SOURCES_DIR
        value: "/home/prow/go/src/github.com/kyma-project/kyma/components/ui-api-layer"

job_labels_template: &job_labels_template
  preset-dind-enabled: "true"
  preset-sa-gcr-push: "true"
  preset-docker-push-repository: "true"

presubmits: # runs on PRs
  kyma-project/kyma:
  - name: &name prow/kyma/components/ui-api-layer
    context: *name
    branches:
    - master
    run_if_changed: "components/ui-api-layer/"
    labels:
      preset-build-pr: "true"
      <<: *job_labels_template
    <<: *job_template
  - name: &name prow/release/kyma/components/ui-api-layer
    context: *name
    run_if_changed: "(components/ui-api-layer/|resources/core/values.yaml)"
    branches:
    - '^release-\d+\.\d+$'
    labels:
      preset-build-release: "true"
      <<: *job_labels_template
    <<: *job_template

```

> ### Calculation release image tag

Reviewer: ```suggestion
### Calculate the release image tag
```

> In previous solution, we used Jenkins for release process. In Jenkins, we can easily specify which version we want to publish, as a job parameter.

Reviewer: ```suggestion
In the internal CI (Jenkins), used for the release process, it is easy to specify which version to publish by adding it as a job parameter.
```

> In Prow, we don't have such option so we are forced to calculate this version. In this proposal, it is achieved by reading git tags. 

Reviewer: ```suggestion
In Prow, such an option is not available. Instead, you need to calculate this version. This proposal describes how to do it by reading Git tags. 
```

> See proposal for export_variables function in prow/scripts/library.sh, where we calculate a new tag.

Reviewer: ```suggestion
See the proposal for the **export_variables** function in `prow/scripts/library.sh`, which shows how to calculate a new tag:
```
